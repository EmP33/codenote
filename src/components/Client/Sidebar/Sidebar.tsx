import React, { useState, useEffect, useCallback } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import styled from "styled-components";

import UserDropdown from "./UserDropdown/UserDropdown";
import UserSettings from "./UserSettings/UserSettings";

import {
  Box,
  Avatar,
  Typography,
  Button,
  ListItem,
  List,
  ListItemButton,
  ListItemText,
  Divider,
  Skeleton,
  Tooltip,
} from "@mui/material";
import SettingsOutlinedIcon from "@mui/icons-material/SettingsOutlined";
import AddIcon from "@mui/icons-material/Add";
import NoteIcon from "@mui/icons-material/Note";
import TaskIcon from "@mui/icons-material/Task";
import HomeIcon from "@mui/icons-material/Home";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

import { useAppSelector } from "../../../lib/hooks";

import { motion, AnimatePresence } from "framer-motion";
import CreateDropdown from "./CreateDropdown/CreateDropdown";

const navVariant = {
  initial: {
    height: "0",
  },
  animate: {
    height: "auto",
  },
  exit: {
    height: "0",
  },
};

const UserDropdownButton = styled.button`
  border: none;
  background: var(--color-secondary-dark);
  color: #fff;
  cursor: pointer;
`;

interface ISidebar {
  user: any;
  onSignOut: () => void;
}

const Sidebar: React.FC<ISidebar> = ({ user, onSignOut }) => {
  const location = useLocation();
  const navigate = useNavigate();
  const userData = useAppSelector((state) => state.user.userData);
  const guestNotes = useAppSelector((state) => state.guest.notes);
  const [showMenu, setShowMenu] = useState(false);
  const [showSettings, setShowSettings] = useState(false);
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);
  const [createAnchor, setCreateAnchor] = useState<HTMLButtonElement | null>(
    null
  );

  useEffect(() => {
    const handleResize = () => {
      if (window.screen.width >= 900) {
        setShowMenu(true);
      } else {
        setShowMenu(false);
      }
    };
    window.addEventListener("resize", handleResize);
    handleResize();
  }, []);

  const expandMenuHandler = () => {
    setShowMenu((prevState) => !prevState);
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = useCallback(() => {
    setAnchorEl(null);
  }, []);

  const openSettingsHandler = () => {
    setShowSettings(true);
  };

  const closeSettingsHandler = useCallback(() => {
    setShowSettings(false);
  }, []);

  const handleCreateClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setCreateAnchor(event.currentTarget);
  };

  const handleCreateClose = useCallback(() => {
    setCreateAnchor(null);
  }, []);

  return (
    <Box
      sx={{
        width: { xs: "100%", md: 300 },
        minHeight: { xs: 0, md: "95vh" },
        height: "auto",
        background: "var(--color-secondary-dark)",
        borderRadius: { xs: 0, md: 2 },
        marginRight: { xs: 0, md: 1 },
        display: "grid",
        gridTemplateRows: "repeat(4,max-content)",
        gridRowGap: 5,
        alignItems: "center",
        position: "relative",
        border: { xs: 0, md: "1px solid var(--color-secondary-dark)" },
      }}
    >
      <Box
        sx={{
          display: "flex",
          alignItems: "center",
          p: 1,
          background: "var(--color-secondary-dark)",
          borderRadius: "8px 8px 0 0 ",
        }}
      >
        {user !== true ? (
          <Avatar
            alt={user?.displayName || user?.email}
            sx={{
              background: "var(--color-tertiary-dark)",
              border: "1px solid #fff",
            }}
          >
            {user
              ? user.displayName
                ? user?.displayName[0]
                : user?.email[0]
              : "G"}
          </Avatar>
        ) : (
          <Skeleton
            variant="circular"
            sx={{ bgcolor: "var(--color-tertiary-dark)" }}
            width={40}
            height={40}
          />
        )}

        <Typography
          variant="body1"
          sx={{ flex: 1, marginLeft: 1, display: "flex" }}
        >
          {user ? (
            user !== true ? (
              user?.displayName || user?.email
            ) : (
              <Skeleton
                sx={{
                  bgcolor: "var(--color-tertiary-dark)",
                  marginRight: 1,
                  width: "100%",
                }}
              />
            )
          ) : (
            "Guest"
          )}

          <UserDropdownButton onClick={handleClick}>
            <ExpandMoreIcon />
          </UserDropdownButton>
        </Typography>
        <Tooltip title="Settings">
          <SettingsOutlinedIcon
            onClick={openSettingsHandler}
            sx={{ "&:hover": { color: "#ccc", cursor: "pointer" } }}
          />
        </Tooltip>
        <UserSettings
          onClose={closeSettingsHandler}
          showSettings={showSettings}
        />
      </Box>
      {user !== true && (
        <UserDropdown
          user={user}
          handleClose={handleClose}
          anchorEl={anchorEl}
          onSignOut={onSignOut}
          open={Boolean(anchorEl)}
        />
      )}
      <Button
        id="signOutButton"
        variant="contained"
        disableElevation
        sx={{
          background: "var(--color-secondary)",

          "&:hover": { background: "var(--color-tertiary-light)" },
        }}
        color="secondary"
        size="large"
        startIcon={<AddIcon />}
        onClick={handleCreateClick}
      >
        Create
      </Button>
      <CreateDropdown
        open={Boolean(createAnchor)}
        anchorEl={createAnchor}
        handleClose={handleCreateClose}
      />
      <Button
        onClick={expandMenuHandler}
        color="inherit"
        sx={{ display: { xs: "block", md: "none" } }}
      >
        <ExpandMoreIcon
          style={{
            margin: "0 auto",
            fontSize: 30,
            transform: showMenu ? "rotate(-180deg)" : "rotate(0)",
            transition: "all .2s ease-in-out",
          }}
        />
      </Button>
      <AnimatePresence>
        {showMenu && (
          <motion.nav
            aria-label="main mailbox folders"
            style={{
              height: "auto",
              overflow: "hidden",
            }}
            variants={navVariant}
            initial="initial"
            exit="exit"
            animate="animate"
          >
            <List>
              <ListItem disablePadding>
                <ListItemButton
                  selected={
                    !location.pathname.includes("notes") &&
                    !location.pathname.includes("tasks")
                  }
                  onClick={() => {
                    location.pathname.includes("client")
                      ? navigate(`/client`)
                      : navigate(`/guest`);
                  }}
                >
                  <HomeIcon sx={{ color: "#fff", marginRight: 2 }} />
                  <ListItemText primary="Home" />
                </ListItemButton>
              </ListItem>
              <Divider />
              <ListItem disablePadding>
                <ListItemButton
                  selected={location.pathname.includes("notes")}
                  onClick={() => {
                    location.pathname.includes("/client")
                      ? userData && userData?.notes
                        ? navigate(
                            //@ts-ignore
                            `/client/notes/${userData.notes[0].id}`
                          )
                        : navigate(`/client/notes/new`)
                      : guestNotes[0]
                      ? navigate(
                          //@ts-ignore
                          `/guest/notes/${guestNotes[0].id}`
                        )
                      : navigate("/guest/notes/new");
                  }}
                >
                  <NoteIcon sx={{ color: "#fff", marginRight: 2 }} />
                  <ListItemText primary="Notes" />
                </ListItemButton>
              </ListItem>
              <Divider />
              <ListItem disablePadding>
                <ListItemButton
                  selected={location.pathname.includes("tasks")}
                  onClick={() => {
                    location.pathname.includes("/client")
                      ? navigate(`/client/tasks`)
                      : navigate(`/guest/tasks`);
                  }}
                >
                  <TaskIcon sx={{ color: "#fff", marginRight: 2 }} />
                  <ListItemText primary="Tasks" />
                </ListItemButton>
              </ListItem>
            </List>
          </motion.nav>
        )}
      </AnimatePresence>

      <Button
        variant="contained"
        sx={{
          display: { xs: "none", md: "block" },
          background: "var(--color-secondary)",
          justifySelf: "center",
          position: "absolute",
          bottom: 20,
          "&:hover": { background: "var(--color-tertiary-dark)" },
        }}
        size="large"
        onClick={onSignOut}
      >
        Sign Out
      </Button>
    </Box>
  );
};

export default Sidebar;
