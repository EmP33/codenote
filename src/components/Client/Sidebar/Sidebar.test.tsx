import { shallow } from "enzyme";
import Sidebar from "./Sidebar";

describe("<Sidebar/>", () => {
  const mockUser = {
    displayName: "Test User",
    email: "test@gmail.com",
    isLoggedIn: true,
  };
  const wrapper = shallow(
    <Sidebar user={mockUser} onSignOut={() => (mockUser.isLoggedIn = false)} />
  );

  it("should render Sidebar component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
