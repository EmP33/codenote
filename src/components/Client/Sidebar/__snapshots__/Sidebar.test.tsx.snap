// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar/> should render Sidebar component 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "alignItems": "center",
      "background": "var(--color-tertiary)",
      "border": Object {
        "md": "1px solid var(--color-tertiary-dark)",
        "xs": 0,
      },
      "borderRadius": Object {
        "md": 2,
        "xs": 0,
      },
      "display": "grid",
      "gridRowGap": 5,
      "gridTemplateRows": "repeat(4,max-content)",
      "height": Object {
        "md": "95vh",
        "xs": "auto",
      },
      "marginRight": Object {
        "md": 1,
        "xs": 0,
      },
      "position": "relative",
      "width": Object {
        "md": 300,
        "xs": "100%",
      },
    }
  }
>
  <ForwardRef(Box)
    sx={
      Object {
        "alignItems": "center",
        "background": "var(--color-tertiary)",
        "borderRadius": "8px 8px 0 0 ",
        "display": "flex",
        "p": 1,
      }
    }
  >
    <ForwardRef(Avatar)
      alt="Test User"
      sx={
        Object {
          "background": "var(--color-tertiary-light)",
          "border": "1px solid #fff",
        }
      }
    >
      T
    </ForwardRef(Avatar)>
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flex": 1,
          "marginLeft": 1,
        }
      }
      variant="body1"
    >
      Test User
      <styled.button
        onClick={[Function]}
      >
        <Memo(ForwardRef(ExpandMoreIcon)) />
      </styled.button>
    </ForwardRef(Typography)>
    <ForwardRef(Tooltip)
      title="Settings"
    >
      <Memo(ForwardRef(SettingsOutlinedIcon))
        sx={
          Object {
            "&:hover": Object {
              "color": "#ccc",
              "cursor": "pointer",
            },
          }
        }
      />
    </ForwardRef(Tooltip)>
  </ForwardRef(Box)>
  <Memo(UserDropdown)
    anchorEl={null}
    handleClose={[Function]}
    onSignOut={[Function]}
    open={false}
    user={
      Object {
        "displayName": "Test User",
        "email": "test@gmail.com",
        "isLoggedIn": true,
      }
    }
  />
  <ForwardRef(Button)
    color="secondary"
    disableElevation={true}
    id="signOutButton"
    size="large"
    startIcon={<Memo />}
    sx={
      Object {
        "&:hover": Object {
          "background": "var(--color-tertiary-dark)",
        },
        "background": "var(--color-tertiary-light)",
      }
    }
    variant="contained"
  >
    Create
  </ForwardRef(Button)>
  <ForwardRef(Button)
    color="inherit"
    onClick={[Function]}
    sx={
      Object {
        "display": Object {
          "md": "none",
          "xs": "block",
        },
      }
    }
  >
    <Memo(ForwardRef(ExpandMoreIcon))
      style={
        Object {
          "fontSize": 30,
          "margin": "0 auto",
        }
      }
    />
  </ForwardRef(Button)>
  <AnimatePresence />
  <ForwardRef(Button)
    color="secondary"
    onClick={[Function]}
    size="large"
    sx={
      Object {
        "&:hover": Object {
          "background": "var(--color-tertiary-dark)",
        },
        "background": "var(--color-tertiary-light)",
        "bottom": 20,
        "display": Object {
          "md": "block",
          "xs": "none",
        },
        "justifySelf": "center",
        "position": "absolute",
      }
    }
    variant="contained"
  >
    Sign Out
  </ForwardRef(Button)>
</ForwardRef(Box)>
`;
